onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -171,6 +171,90 @@ activation.)DOC")
       })
       .SetDoc(R"DOC(Tokenizer divides each string in X into a vector of strings along the last axis. All input strings including attributes are UTF-8 encoded.)DOC");
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Ngram)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "X", "Input for n-gram extraction", "T")
+      .Output(0, "Y", "NGram results", "T1")
+      .TypeConstraint(
+          "T",
+          {"tensor(string)", "tensor(int32)", "tensor(int64)"},
+          "Input is ether string UTF-8 or int32/int64")
+      .TypeConstraint(
+          "T1",
+          {"tensor(float)"},
+          "1-D tensor of floats")
+      .Attr(
+          "N",
+          "Maximum n-gram length",
+          AttributeProto::INT)
+      .Attr(
+          "M",
+          "Minimum n-gram length",
+          AttributeProto::INT)
+      .Attr(
+          "all",
```

Comment:
>all [](start = 11, length = 3)

Why is this needed? If you just want N, why not have the same value for M and N?

Reply:
My guess is that this operator is modeled after the existing operators that are already in existence/use by ML .NET and so people want to see the familiar names. @wschin@outlook.com 

---
In reply to: [243103334](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243103334) [](ancestors = 243103334)

Reply:
Replied to the same comment in ngram_test.cc. [Copied reply] You're right. We can safely remove all. all=true will be replaced with M=1 and N=N. all=false can be encoded as M=N.


---
In reply to: [243110192](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243110192) [](ancestors = 243110192,243103334)